### PR TITLE
Fix bug #1435 (data loss when dumping payloads to JSON)

### PR DESCRIPTION
**Pull request by gozzy**
This pull request tries to fix bug #1435. Data should not be lost any more.

### CODE REVIEW

**Comment by gozzy** [comment]
Also this version should work a bit faster (because of memmove), at least when 'payload_printable' is set to 'no'.

File: src/output-json-alert.c [hunk 1]
```diff
@@ -93,12 +93,13 @@ typedef struct JsonAlertLogThread_ {
...
  * so we can report them in JSON output. */
-static int AlertJsonPrintStreamSegmentCallback(const Packet *p, void *data, uint8_t *buf, uint32_t buflen)
+static int AlertJsonDumpStreamSegmentCallback(const Packet *p, void *data, uint8_t *buf, uint32_t buflen)
 {
     MemBuffer *payload = (MemBuffer *)data;
 
-    PrintStringsToBuffer(payload->buffer, &payload->offset, payload->size,
-                         buf, buflen);
+    // We can't user Print...ToBuffer functions as we need real raw data.
+    memmove(payload->buffer + payload->offset, buf, buflen);
```
Comment (inliniac) [inline src/output-json-alert.c]: If I understand memmove correctly, it is a memcpy with a intermediate step: "The  memmove()  function  copies  n bytes from memory area src to memory area dest.  The memory areas may overlap: copying takes place as though the bytes in src are first copied into a temporary array that does not overlap src or dest, and the bytes are then copied from the temporary array to dest." (from the man page)

Here we don't seem to have an overlap case, so then memcpy would make more sense?

File: src/output-json-alert.c [hunk 2]
```diff
@@ -287,27 +288,30 @@ static int AlertJson(ThreadVars *tv, JsonAlertLogThread *aft, const Packet *p)
                 }
 
                 if (json_output_ctx->flags & LOG_JSON_PAYLOAD) {
+                    unsigned char printable_buf[payload->offset + 1];
```
Comment (inliniac) [inline src/output-json-alert.c]: I prefer using types like uint8_t here

File: src/output-json-alert.c [hunk 3]
```diff
@@ -287,27 +288,30 @@ static int AlertJson(ThreadVars *tv, JsonAlertLogThread *aft, const Packet *p)
...
-                unsigned char packet_buf[p->payload_len + 1];
-                uint32_t offset = 0;
-
-                PrintStringsToBuffer(packet_buf, &offset,
-                                     p->payload_len + 1,
-                                     p->payload, p->payload_len);
-
                 if (json_output_ctx->flags & LOG_JSON_PAYLOAD_BASE64) {
-                    unsigned long len = sizeof(packet_buf) * 2;
+                    unsigned long len = p->payload_len * 2 + 1;
```
Comment (inliniac) [inline src/output-json-alert.c]: can we use a more explicit type here? I always have to think about the size of long

**Comment by gozzy** [comment]
Replaced with #1455

**Pull request closed without merging**